### PR TITLE
expression: fix builtin function `ord`

### LINKED ISSUE #13993: Incorrect behavior of `ord` when input is binary
## Bug Report

Please answer these questions before submitting your issue. Thanks!

1. What did you do?

Give ord a binary input, see https://github.com/tikv/tikv/pull/6189#issuecomment-563090152

2. What did you expect to see?

```
mysql> select ord(X'e4bda0');
+----------------+
| ord(X'e4bda0') |
+----------------+
|            228 |
+----------------+
1 row in set (0.00 sec)
```

3. What did you see instead?

```
TiDB> select ord(X'e4bda0');
+----------------+
| ord(X'e4bda0') | 
+----------------+
| 14990752       | 
+----------------+
1 row in set (0.00 sec)
```

4. What version of TiDB are you using (`tidb-server -V` or run `select tidb_version();` on TiDB)?

```
Release Version: v4.0.0-alpha-677-g18e42cfd2
Git Commit Hash: 18e42cfd233fa04850c24469c413237e38f02726
Git Branch: feature/wasm
UTC Build Time: 2019-11-28 08:41:07
GoVersion: go version go1.13.4 linux/amd64
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false
```

### PR DESCRIPTION
**Pull request by Aloxaf**
Signed-off-by: Aloxaf <aloxafx@gmail.com>

### What problem does this PR solve? 

Fix #13993 

### What is changed and how it works?

Split `ord` into `ordUtf8` and `ordBinary` and choose the correct one according to charset

### Check List 

Tests 

 - Integration test

Code changes

No

Side effects

No

Related changes

No

Release note

No

### CODE REVIEW

**Comment by Aloxaf** [comment]
Ummmm, no just binary.

MySQL:
```
mysql> select ord(_ascii'你');
+------------------+
| ord(_ascii'你')  |
+------------------+
|              228 |
+------------------+
1 row in set, 1 warning (0.00 sec)
```

TiDB:
```
mysql> select ord(_ascii'你');
+------------------+
| ord(_ascii'你')  |
+------------------+
|         14990752 |
+------------------+
1 row in set (0.00 sec)
```

I think I should look into the MySQL's source code.

**Comment by Aloxaf** [comment]
According to [MySQL's implementation](https://github.com/mysql/mysql-server/blob/8.0/sql/item_func.cc#L3713-L3730), `ord` will check if the charset has an `ismbchar` method, if so it will be called to extract a character, otherwise return the first byte.

The implementation of `ismbchar` can be found if `strings/ctype-{CHARSET}.cc` (eg. [ctype-utf8.cc](https://github.com/mysql/mysql-server/blob/8.0/strings/ctype-utf8.cc#L5964)).
According to these, in the charset supported by TiDB, `ascii`, `binary` and `latin1` should return the first byte.

**Comment by Deardrops** [comment]
LGTM

File: expression/builtin_string.go [hunk 1]
```diff
@@ -2701,10 +2701,31 @@ func (b *builtinOrdSig) evalInt(row chunk.Row) (int64, bool, error) {
...
+	desc, err := charset.GetCharsetDesc(charSet)
+	if err != nil {
+		return 0, isNull, err
+	}
+
+	if desc.Maxlen == 1 {
+		return ordBinary(str), false, nil
+	}
+	return ordUtf8(str), false, nil
+}
```
Comment (wjhuang2016) [inline expression/builtin_string.go]: Please extract these code as a function name by `chooseOrdFunc` or other. Then we can reuse the function in vecEvalInt.

File: expression/builtin_string.go [hunk 2]
```diff
@@ -2701,10 +2701,31 @@ func (b *builtinOrdSig) evalInt(row chunk.Row) (int64, bool, error) {
...
+
+func ordBinary(str string) int64 {
```
Comment (wjhuang2016) [inline expression/builtin_string.go]: s/ordBinay/ordSingleByte

Comment (Aloxaf) [reply expression/builtin_string.go]: Renamed

File: expression/builtin_string.go (hunk 1 above)
Comment (Aloxaf) [reply expression/builtin_string.go]: Finished

**Pull request merged**